File: 11_bridge/java/Bridge.java
public class Bridge {
    public static void main(String[] args) {
        PaymentProcess visaPaymentProcessor = new VisaPaymentProcess("1234.3456.xxx", "12/25", "123");
        MemberRegistration membership = new MemberRegistration(visaPaymentProcessor);
        membership.register();

        PaymentProcess momoPaymentProcessor = new MomoPaymentProcess("09090909090");
        MemberRegistration membershipMomo = new MemberRegistration(momoPaymentProcessor);
        membershipMomo.register();
    }
}

abstract class PaymentProcess {
    public abstract void pay(double amount);
}

class VisaPaymentProcess extends PaymentProcess {
    private final String cardNumber;
    private final String expiryDate;
    private final String cvv;

    public VisaPaymentProcess(String cardNumber, String expiryDate, String cvv) {
        this.cardNumber = cardNumber;
        this.expiryDate = expiryDate;
        this.cvv = cvv;
    }

    @Override
    public void pay(double amount) {
        System.out.println("Paying " + amount + " USD with visa card " + cardNumber + "...");
        // TODO: Implement logic...
    }
}

class MomoPaymentProcess extends PaymentProcess {
    private final String phoneNumber;

    public MomoPaymentProcess(String phoneNumber) {
        this.phoneNumber = phoneNumber;
    }

    @Override
    public void pay(double amount) {
        System.out.println("Paying " + amount + " VND with momo card " + phoneNumber + "...");
        // TODO: Implement logic...
    }
}

class MemberRegistration {
    private final PaymentProcess paymentProcessor;

    public MemberRegistration(PaymentProcess paymentProcessor) {
        this.paymentProcessor = paymentProcessor;
    }

    public void register() {
        double amount = 100; // the registration fee
        paymentProcessor.pay(amount);
        System.out.println("Registered for Youtube membership!");
    }
}

File: 11_bridge/javascript/bridge.js
// define PaymentProcess
class PaymentProcess {
  pay(amount) {
  }
}

// define VisaPaymentProcess class
class VisaPaymentProcess extends PaymentProcess {
  constructor(cardNumber, expiryDate, cvv){
    super()
    this.cardNumber = cardNumber
    this.expiryDate = expiryDate
    this.cvv = cvv
  }

  // implements the pay method
  pay(amount){
    console.log(`Paying ${amount} USD with visa card ${this.cardNumber}...`)
    // TODO: Implement logic...
  }
}

// define MomoPaymentProcess
class MomoPaymentProcess extends PaymentProcess {
  constructor(phoneNumber){
    super()
    this.phoneNumber = phoneNumber
  }

  // implements the pay method
  pay(amount){
    console.log(`Paying ${amount} VND with momo card ${this.phoneNumber}...`)
    // TODO: Implement logic...
  }
}


// define MemberRegistration
class MemberRegistration {
  constructor(paymentProcessor) {
    this.paymentProcessor = paymentProcessor
  }

  // register
  register(){
    const amount = 100; // the registration fee in USD
    this.paymentProcessor.pay(amount)
    console.log('Registered for Youtube membership!')
  }
}

// Create visa payment
const visaPaymentProcessor = new VisaPaymentProcess('1234.3456.xxx', '12/25', '123')
const membership = new MemberRegistration(visaPaymentProcessor)
membership.register()

// create momo payment
const momoPaymentProcessor = new MomoPaymentProcess('09090909090')
const membershipMomo = new MemberRegistration(momoPaymentProcessor)
membershipMomo.register()